query-parser/src/parse.rs

```diff
@@ -83,320 +96,348 @@ error_chain! {
+    choice::<[&mut Parser<Input = _, Output = FindSpec>; 4], _>
         ([&mut try(Find::find_scalar()),
           &mut try(Find::find_coll()),
           &mut try(Find::find_tuple()),
           &mut try(Find::find_rel())])
-        .parse_stream(input)
 });
 
-// Parse a sequence of values into one of four find specs.
-//
-// `:find` must be an array of plain var symbols (?foo), pull expressions, and aggregates.
-// For now we only support variables and the annotations necessary to declare which
-// flavor of :find we want:
-//
-//
-//     `?x ?y ?z  `     = FindRel
-//     `[?x ...]  `     = FindColl
-//     `?x .      `     = FindScalar
-//     `[?x ?y ?z]`     = FindTuple
-//
-#[allow(dead_code)]
-pub fn find_seq_to_find_spec(find: &[edn::Value]) -> FindParseResult {
-    Find::find()
-        .parse(find)
-        .map(|x| x.0)
-        .map_err::<ValueParseError, _>(|e| e.translate_position(find).into())
-        .map_err(|e| Error::from_kind(ErrorKind::FindParseError(e)))
+/// TODO: extract macro for matching these `Keyword` instances.
+def_parser!(Find, literal_find, edn::ValueAndSpan, {
+    satisfy(|v: edn::ValueAndSpan| {
+        if let edn::SpannedValue::Keyword(ref s) = v.inner {
+            s.0.as_str() == "find"
+        } else {
+            false
+        }
+    })
+});
+
+def_parser!(Find, literal_with, edn::ValueAndSpan, {
+    satisfy(|v: edn::ValueAndSpan| {
+        if let edn::SpannedValue::Keyword(ref s) = v.inner {
+            s.0.as_str() == "with"
+        } else {
+            false
+        }
+    })
+});
+
+def_parser!(Find, literal_where, edn::ValueAndSpan, {
+    satisfy(|v: edn::ValueAndSpan| {
+        if let edn::SpannedValue::Keyword(ref s) = v.inner {
+            s.0.as_str() == "where"
+        } else {
+            false
+        }
+    })
+});
+
+/// Express something close to a builder pattern for a `FindQuery`.
+enum FindQueryPart {
+    FindSpec(FindSpec),
+    With(Vec<Variable>),
+    WhereClauses(Vec<WhereClause>),
 }
 
-#[allow(dead_code)]
-pub fn clause_seq_to_patterns(clauses: &[edn::Value]) -> WhereParseResult {
-    Where::clauses()
-        .parse(clauses)
+/// This is awkward, but will do for now.  We use `keyword_map()` to optionally accept vector find
+/// queries, then we use `FindQueryPart` to collect parts that have heterogeneous types; and then we
+/// construct a `FindQuery` from them.
+def_parser!(Find, query, FindQuery, {
+    let p_find_spec = Find::literal_find()
+        .with(vector().of(Find::spec().map(FindQueryPart::FindSpec)));
+
+    let p_with_vars = Find::literal_with()
+        .with(vector().of(many(Query::variable()).map(FindQueryPart::With)));
+
+    let p_where_clauses = Find::literal_where()
+        .with(vector().of(Where::clauses().map(FindQueryPart::WhereClauses))).expected(":where clauses");
+
+    (or(map(), keyword_map()))
+        .of(many(choice::<[&mut Parser<Input = ValueStream, Output = FindQueryPart>; 3], _>([
```

Comment:
Hm.

I have to say, the components of this parser do not build up to be used like this; this looks like it's inefficient and produces worse error messages.

`or(map(), keyword_map())` is acceptable, I think, but beyond that I wouldn't use `choice`.

I think I'd write a `satisfy_map` parser that consumed the output of the `or`, programmatically looked at the expected keys, and processed the values accordingly using three individual parsers. Here you've got a much more expensive parse by using `choice` — it'll typically try `p_find_spec` four times, once for `:find`, once for `:in`, once for `:with`, and once for `:where`, then `p_with_vars` three times, and so on.

You're also losing the opportunity for the precisely typed, useful error messages that come out of the code you deleted, replacing them with whatever `combine` provides as an error when given this tree of parsers.

Given input like `{:in [] :where [[?x _ ?y]]}` the old code returned `ErrorKind::MissingFieldError(:find)`.

I believe this code now says "unexpected token: expected :find at position -2123981209243", or something like that. No span is meaningful here, and it's not an unexpected token…

Can we do better by accepting that this is not well phrased as a `combine` parser?